File: frontend/src/constants/tables.ts
/**
 * System Table Constants
 * 
 * Imports from shared JSON - single source of truth matching protobuf enums.
 * @see shared/constants/system.json
 * @see shared/proto/constants.proto (SystemTable enum)
 */

import systemConstants from '@shared/constants/system.json';

// Extract system tables from shared JSON
const { systemTables } = systemConstants;

// Table names - derived from shared JSON
export const TableObject = systemTables.OBJECT;
export const TableField = systemTables.FIELD;
export const TableRelationship = systemTables.RELATIONSHIP;
export const TableRecordType = systemTables.RECORD_TYPE;
export const TableLayout = systemTables.LAYOUT;
export const TableApp = systemTables.APP;
export const TableDashboard = systemTables.DASHBOARD;
export const TableListView = systemTables.LIST_VIEW;
export const TableSetupPage = systemTables.SETUP_PAGE;
export const TableProfile = systemTables.PROFILE;
export const TableUser = systemTables.USER;
export const TableObjectPerms = systemTables.OBJECT_PERMS;
export const TableFieldPerms = systemTables.FIELD_PERMS;
export const TableFlow = systemTables.FLOW;
export const TableAction = systemTables.ACTION;
export const TableValidation = systemTables.VALIDATION;
export const TableRecycleBin = systemTables.RECYCLE_BIN;
export const TableLog = systemTables.LOG;
export const TableRecent = systemTables.RECENT;
export const TableConfig = systemTables.CONFIG;
export const TableUIComponent = systemTables.UI_COMPONENT;

/**
 * Check if a table name is a system table
 */
export function isSystemTable(tableName: string): boolean {
    return tableName.startsWith('_System_');
}

File: backend/pkg/constants/tables.go
package constants

import "strings"

// System table names - used throughout the codebase for metadata and system operations.
// These are the internal tables that store platform configuration.
const (
	// Core Metadata Tables
	SystemTablePrefix = "_System_"
	// TableObject, TableField, TableRelationship, TableRecordType defined in z_generated.go
	TableAutoNumber              = "_System_AutoNumber"
	TableTable                   = "_System_Table"
	TableGroup                   = "_System_Group"
	TableGroupMember             = "_System_GroupMember"
	TablePermissionSet           = "_System_PermissionSet"
	TablePermissionSetAssignment = "_System_PermissionSetAssignment"

	// UI & Layout Tables
	// TableLayout defined in z_generated.go
	// TableApp defined in z_generated.go
	TableTheme = "_System_Theme"
	// TableDashboard, TableListView, TableUIComponent, TableSetupPage defined in z_generated.go

	// Security & Permissions Tables
	// TableProfile, TableUser, TableRole defined in z_generated.go
	TableSession = "_System_Session"
	// TableObjectPerms, TableFieldPerms, TableSharingRule defined in z_generated.go
	TableRecordShare = "_System_RecordShare"
	TableTeamMember  = "_System_TeamMember"
	// TableProfileLayout defined in z_generated.go

	// Automation Tables
	// TableFlow defined in z_generated.go
	TableFlowStep     = "_System_FlowStep"
	TableFlowInstance = "_System_FlowInstance"
	// TableAction, TableValidation defined in z_generated.go
	TableFieldDependency  = "_System_FieldDependency"
	TableApprovalProcess  = "_System_ApprovalProcess"
	TableApprovalWorkItem = "_System_ApprovalWorkItem"

	// System Operation Tables
	// TableRecycleBin, TableLog, TableRecent, TableConfig defined in z_generated.go
	TableComment      = "_System_Comment"
	TableNotification = "_System_Notification"
	TableAuditLog     = "_System_AuditLog"
	TableOutboxEvent  = "_System_OutboxEvent"
)

func IsSystemTable(tableName string) bool {
	return strings.HasPrefix(tableName, SystemTablePrefix)
}

// GetSystemFieldNames returns the list of system fields that every table should have.
// Note: id is NOT required since some tables use different primary keys (key_name, composite keys).
func GetSystemFieldNames() []string {
	// Only temporal fields are universally required for operation tracking
	return []string{
		"created_date",
		"last_modified_date",
	}
}

// GetStandardSystemFieldNames returns the full set of system fields for standard/custom objects
func GetStandardSystemFieldNames() []string {
	return []string{
		"id",
		"created_date",
		"last_modified_date",
		"created_by_id",
		"last_modified_by_id",
		"owner_id",
		"is_deleted",
	}
}
